test/cluster/object_store/test_backup.py

```diff
@@ -617,60 +626,127 @@ async def create_dataset(manager, ks, cf, topology, logger, rf_across_dc_boundar
+                assert target_host_id in servers_by_host_id.keys()
+            elif scope == 'dc':
+                assert servers_by_host_id[target_host_id].datacenter == s.datacenter
+            elif scope == 'rack':
+                assert servers_by_host_id[target_host_id].datacenter == s.datacenter
+                assert servers_by_host_id[target_host_id].rack == s.rack
+            elif scope == 'node':
+                assert target_host_id == str(host_ids[s.server_id])
+            streamed_to[target_host_id] += 1
+
+        # validate balance only when rf == #racks
+        if topology.rf != topology.racks:
+            logger.info(f'Skipping balance checks since rf != racks ({topology.rf} != {topology.racks})')
+            continue
+
+        if scope == 'all':
+            assert set(streamed_to.keys()).issubset(set(host_ids.values()))
+            assert len(streamed_to) == topology.rf * topology.dcs
         elif scope == 'dc':
-            scope_nodes.update([ str(host_ids[s.server_id]) for s in servers[i::topology.dcs] ])
-        logger.info(f'{s.ip_addr} streamed to {streamed_to}, expected {scope_nodes}')
-        assert streamed_to == scope_nodes
-
-async def do_restore(ks, cf, s, toc_names, scope, prefix, object_storage, manager, logger, primary_replica_only = False):
+            # it's guaranteed the node replicated only within the datacenter by asserts above
+            assert set(streamed_to.keys()).issubset(set(host_ids_per_dc[s.datacenter]))
+            assert len(streamed_to) == topology.rf
+        elif scope == 'rack' and topology.rf == topology.racks:
+            assert set(streamed_to.keys()).issubset(set(host_ids_per_dc_rack[s.datacenter][s.rack]))
+            assert len(streamed_to) == 1
+
+        # asses balance
+        streamed_to_counts = streamed_to.values()
+        assert len(streamed_to_counts) > 0
+        mean_count = statistics.mean(streamed_to_counts)
+        max_deviation = max(abs(count - mean_count) for count in streamed_to_counts)
+        if not primary_replica_only:
+            assert max_deviation == 0, f'if primary_replica_only is False, streaming should be perfectly balanced: {streamed_to}'
+            continue
+
+        assert max_deviation < 0.1 * mean_count, f'node {s.ip_addr} streaming to primary replicas was unbalanced: {streamed_to}'
+
+async def do_restore_server(ks, cf, s, toc_names, scope, prefix, object_storage, manager, logger, primary_replica_only):
     logger.info(f'Restore {s.ip_addr} with {toc_names}, scope={scope}')
     tid = await manager.api.restore(s.ip_addr, ks, cf, object_storage.address, object_storage.bucket_name, prefix, toc_names, scope, primary_replica_only=primary_replica_only)
     status = await manager.api.wait_task(s.ip_addr, tid)
     assert (status is not None) and (status['state'] == 'done')
 
+async def do_restore(ks, cf, servers, topology, sstables, scope, prefix, object_storage, manager, logger, primary_replica_only = False):
```

Comment:
I think/hope that adding `do_restore_server` below `do_restore` would reduce the diff and make it clearer